File: src/Scripting/Perl/adapter.pl
use strict;
use warnings;

package LegacyFahrenheitSensor;
sub new { bless {}, shift }
sub read_fahrenheit { return 86 }

package FahrenheitSensorAdapter;
sub new {
    my ($class, $legacy) = @_;
    return bless { legacy => $legacy }, $class;
}
sub read_celsius {
    my ($self) = @_;
    my $fahrenheit = $self->{legacy}->read_fahrenheit();
    return int((($fahrenheit - 32) * 5 / 9) + 0.5);
}

package main;
my $legacy = LegacyFahrenheitSensor->new();
my $reader = FahrenheitSensorAdapter->new($legacy);
print 'legacy=' . $legacy->read_fahrenheit() . "F\n";
print 'adapted=' . $reader->read_celsius() . "C\n";

File: src/Web/Dart/adapter.dart
abstract interface class TemperatureReader {
  int readCelsius();
}

final class LegacyFahrenheitSensor {
  int readFahrenheit() => 86;
}

final class FahrenheitSensorAdapter implements TemperatureReader {
  FahrenheitSensorAdapter(this._adaptee);

  final LegacyFahrenheitSensor _adaptee;

  @override
  int readCelsius() => ((_adaptee.readFahrenheit() - 32) * 5) ~/ 9;
}

void main() {
  final legacy = LegacyFahrenheitSensor();
  final TemperatureReader reader = FahrenheitSensorAdapter(legacy);

  print('legacy=${legacy.readFahrenheit()}F');
  print('adapted=${reader.readCelsius()}C');
}
